### DISCUSSION TITLE
app component structure, 'view' components with shadowdom disabled

### DISCUSSION THREAD

**Discussion by ajmchambers** [lit/lit#3081] [General]
It seems to be frowned upon to disable shadowdom in lit components and potentially unsupported in certain scenarios (e.g.: https://github.com/lit/lit/issues/3080).

There was a discussion on slack a little while back about 'view' components:
https://lit-and-friends.slack.com/archives/C03PF4L4L/p1655216297320959

Similar to that thread, I've been finding it useful to disable shadowdom in certain 'view' components. 

Here is an example:

```html
<app-root>
  <app-frame>
    <page-home>
      <ol>
        <li><a href="#section-one">Section One</a></li>
        <li><a href="#section-two">Section Two</a></li>
      </ol>
      <h2 id="section-one">Section One</h2>
      <p>...</p>
      <h2 id="section-two">Section Two</h2>
      <p>...</p>
      <div class="global-styles"></div>
    </page-home>
  </app-frame>
</app-root>
```

- app-root:
  - shadow-dom
  - global state controllers - auth, theme-preference
  - configures router
  - router renders into slot
- app-frame:
  - shadow-dom
  - app navigation menu, user profile, notification drawer
  - global state - eg. notification-count
  - slot for the page component
- page-one:
  - 'view' component, shadowroot disabled
  - affected by global styles
  - anchor links scroll to headings
  - may still have state and use component lifecycle e.g.
    - fetch and render info from database/state management
    - listen to changes
    - update state based on user input changes

NOTE: the app-frame is not always rendered, e.g user has to be logged in.

```html
<app-root>
  <page-login>
    <form>
      ...
    </form>
  </page-login>
</app-root>
```

The routes could look like this:

```javascript
[
  {
    path: '/',
    component: 'app-frame',
    children: [
      {
        path: '/',
        component: 'page-home'
      },
      {
        path: '/todos',
        component: 'page-todos'
      }
    ]
  },
  {
    path: '/login',
    component: 'page-login'
  }
]
```

I prefer this over having the entire app being contained within the root components shadowdom:

```html
<app-root></app-root>
```

This could be partially solved by a router if you could define routes like this (and have it render into the current element's slot):

```javascript
function PageHome() {
  return html`
    <app-frame>
      <div class="page-home">
        <ol>
          <li><a href="#section-one">Section One</a></li>
          <li><a href="#section-two">Section Two</a></li>
        </ol>
        <h2 id="section-one">Section One</h2>
        <p>...</p>
        <h2 id="section-two">Section Two</h2>
        <p>...</p>
        <div class="global-styles"></div>
      </div>
    </app-frame>
  `
}

const routes = [
  {
    path: '/'
    render: () => PageHome()
  }
]

// and in the app-root component, tell it to render the routes into it's own slot.
```

But you would lose the state and component lifecycle from having a 'page-home' component.

Is this unnecessarily fighting with how lit is designed to work here? If the goal was to have the app structured in this way is it better to e.g. use lit for the components - then assemble the components into an app and handle routing and the 'view' components in some other framework?

**Comment by thescientist13** [comment]
Yeah, I've had similar experiences wanting to adopt WebComponents and having reasons where I didn't need / want Shadow DOM and I know as part of my participation in the Web Component Community Group, [this topic](https://github.com/webcomponents-cg/docs-and-guides/issues/21#issuecomment-1134750145) comes up and recommendations seem somewhat split on if Shadow DOM should be "optional" or not.

One case I personally experienced is like yours, where I wasn't ready to refactor all my styles and site design to support Shadow DOM (yet) but at least wanted to migrate my SPA to Lit, coming from an existing a framework.  The router I had used is based on Lit and so by default encapsulated all that view content inside a Shadow Root (e.g. `<lit-route>`), which meant [all my global bootstrap styles and FontAwesome icons were never reflected in the page](https://github.com/AnalogStudiosRI/www.analogstudios.net/pull/56), and thus I had to include the styles as CSS-in-JS for each component which really ballooned up the bundle size.   That said, on the client side, assuming you own your router (which I did not), with Lit it should be easy as using `createRenderRoot` to escape this as needed.

However, in the case of #3080, I think it would be awesome to leverage Web Components as an alternative to 3rd party templating formats (handlebars, liquid, etc.) to generate [page level HTML in SSG / SSR contexts, a la Next](https://twitter.com/thegreenhouseio/status/1540070019383664640) but following a familiar and predictable browser standard.  This is what I was able to accomplish with a simple native `HTMLElement` and works great for my project [WCC](https://github.com/ProjectEvergreen/wcc) which I am in the process of [integrating into my own project Greenwood](https://github.com/ProjectEvergreen/greenwood/pull/943) to support page and layout authoring like so.
```js
// src/pages/artists.js
import fetch from 'node-fetch';
import '../components/card/card.js';  // <wc-card></wc-card>

export default class ArtistsPage extends HTMLElement {
  async connectedCallback() {
    const artists = await fetch('https://www.example.com/api/artists').then(resp => resp.json());
    const html = artists(artist => {
      return `
        <wc-card>
          <h2 slot="title">${artist}</h2>
          <img slot="image" src="${artist}" alt="${artist}"/>
        </wc-card>
      `;
    }).join('');

    this.innerHTML = html;
  }
}
```

I wanted to do the same for Lit+SSR but unfortunately it seems that `createRenderRoot` will not be supported server-side so thus SSR users of Lit who want to SSR pages / page layouts like in the above example will have to:
1. Depend on a browser supporting Declarative Shadow DOM natively (which is only Chrome / Edge right now)
1. Load the JavaScript polyfill for those other browsers, defeating the whole purpose of trying to ship static HTML / content by default

In addition, I'm not sure if [search engines and web crawlers can detect content inside a `<template>` tag](https://lit-and-friends.slack.com/archives/C03PF4L4L/p1649882412643339)?
